nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java

```diff
@@ -0,0 +1,216 @@
+    }
+
+    @Override
+    public final void onConfigured(final LoginIdentityProviderConfigurationContext configurationContext) throws ProviderCreationException {
+        final Map<String, String> configProperties = configurationContext.getProperties();
+        for (String propertyKey : configProperties.keySet()) {
+            String propValue = configProperties.get(propertyKey);
+            logger.debug("found property '{}': '{}'", propertyKey, propValue);
+        }
+
+        credentialsFilePath = configProperties.get(PROPERTY_CREDENTIALS_FILE);
+        if (credentialsFilePath == null) {
+            final String message = String.format("Identity Provider '%s' requires a credentials file path in property '%s'",
+                    identifier, PROPERTY_CREDENTIALS_FILE);
+            throw new ProviderCreationException(message);
+        }
+
+        final String rawExpirationPeriod = configProperties.get(PROPERTY_EXPIRATION_PERIOD);
+        if (rawExpirationPeriod == null || rawExpirationPeriod.isEmpty()) {
+            final String message = String.format("Identity Provider '%s' requires a credential expiration in property '%s'",
+                    identifier, PROPERTY_EXPIRATION_PERIOD);
+            throw new ProviderCreationException(message);
+        } else {
+            try {
+                expirationPeriodMilliseconds = FormatUtils.getTimeDuration(rawExpirationPeriod, TimeUnit.MILLISECONDS);
+            } catch (IllegalArgumentException iae) {
+                final String message = String.format("Identity Provider '%s' property '%s' value of '%s', is not a valid time period",
+                        identifier, PROPERTY_EXPIRATION_PERIOD, rawExpirationPeriod);
+                throw new ProviderCreationException(message);
+            }
+        }
+
+        logger.debug("Identity Provider '{}' configured to use file '{}' and expiration period of '{}'={} milliseconds",
+                identifier, credentialsFilePath, rawExpirationPeriod, expirationPeriodMilliseconds);
+    }
+
+    String getCredentialsFilePath() {
+        return credentialsFilePath;
+    }
+
+    long getExpirationPeriod() {
+        return expirationPeriodMilliseconds;
+    }
+
+    @Override
+    public final AuthenticationResponse authenticate(final LoginCredentials credentials) throws InvalidLoginCredentialsException, IdentityAccessException {
+        final String loginUsername = credentials.getUsername();
+        final String loginPassword = credentials.getPassword();
+        AuthenticationResponse authResponse = null;
+
+        try {
+            UserCredentialsList credentialsList = loadCredentialsList(credentialsFilePath);
+            List<UserCredentials> userCredentials = credentialsList.getUser();
+            for (UserCredentials userCreds : userCredentials) {
+                final String storedUsername = userCreds.getName();
+                if (storedUsername.equals(loginUsername)) {
```

Comment:
I don't see anywhere that a user can register, so the credential store on the filesystem can only currently be modified manually by someone with `write` access to that file. However, it doesn't appear that there are any safeguards on the data contained therein -- no unique constraint on the usernames, so in the following scenario, what would happen?

```xml
<credentials>
    <user name="user1" passwordHash="bcrypt(password)" />
    <user name="user1" passwordHash="bcrypt(real&Strong&Password)" />
</credentials>
```

If a login request for `user1`, `password` came in, would it succeed? What about `user1`, `real&Strong&Password`?

What is the mechanism for an administrator adding a user to the credential store? They must write a bespoke Java method to hash the user-provided password and put the resulting hash into the XML file by hand? This is the same for password resets and updates?

There is no default  `credentials.xml` file provided with this NAR but should an empty one be provided with default file system permissions locked down to further indicate that anyone with `read` access to this file can perform brute-force or pattern-matching hash analysis against the credentials store, and anyone with `write` access can simply replace a user's password hash and then gain access to the system with their account?

I understand this is not designed for a production system but it still seems dangerous to release as is, because well-intentioned but short-on-time people will deploy this to production.